File: ios/PttTalk/Sources/PttTalkLib/EncryptedChat.swift
import CryptoKit
import Foundation

public enum ChatContentKind: UInt8, Codable, CaseIterable, Sendable {
    case text = 1
    case file = 2
    case voice = 3
    case video = 4
}

public struct ChatAttachment: Codable, Equatable, Sendable {
    public let attachmentId: UUID
    public let fileName: String
    public let mimeType: String
    public let plaintextBytes: Int64
    public let durationMs: Int32
    public let key: Data
    public let ciphertextSha256: Data

    public init(
        attachmentId: UUID,
        fileName: String,
        mimeType: String,
        plaintextBytes: Int64,
        durationMs: Int32 = 0,
        key: Data,
        ciphertextSha256: Data
    ) {
        self.attachmentId = attachmentId
        self.fileName = fileName
        self.mimeType = mimeType
        self.plaintextBytes = plaintextBytes
        self.durationMs = durationMs
        self.key = key
        self.ciphertextSha256 = ciphertextSha256
    }
}

public struct ChatMessage: Codable, Equatable, Identifiable, Sendable {
    public let messageId: UUID
    public let channelId: UUID
    public let membershipEpoch: Int32
    public let sentAt: Date
    public let senderAci: String
    public let senderDeviceId: Int
    public let kind: ChatContentKind
    public let text: String
    public let attachment: ChatAttachment?
    public var id: UUID { messageId }

    public init(
        messageId: UUID,
        channelId: UUID,
        membershipEpoch: Int32,
        sentAt: Date,
        senderAci: String,
        senderDeviceId: Int,
        kind: ChatContentKind,
        text: String,
        attachment: ChatAttachment? = nil
    ) {
        self.messageId = messageId
        self.channelId = channelId
        self.membershipEpoch = membershipEpoch
        self.sentAt = sentAt
        self.senderAci = senderAci
        self.senderDeviceId = senderDeviceId
        self.kind = kind
        self.text = text
        self.attachment = attachment
    }
}

public enum EncryptedChatCodec {
    private static let magic = Data("PTTC".utf8)
    private static let attachmentMagic = Data("PTTA".utf8)
    public static let maximumTextBytes = 4_096
    public static let maximumAttachmentBytes = 25 * 1_024 * 1_024

    static func boundedUTF8(_ value: String, maximumBytes: Int) -> String {
        guard Data(value.utf8).count > maximumBytes else { return value }
        var result = ""
        for character in value {
            guard Data((result + String(character)).utf8).count <= maximumBytes else { break }
            result.append(character)
        }
        return result
    }

    public static func encode(_ message: ChatMessage) throws -> Data {
        guard message.membershipEpoch > 0, (1...2).contains(message.senderDeviceId),
              UUID(uuidString: message.senderAci) != nil else { throw EncryptedChatError.invalidMessage }
        let text = Data(message.text.utf8)
        guard text.count <= maximumTextBytes else { throw EncryptedChatError.textTooLarge }
        if message.kind == .text, message.attachment != nil { throw EncryptedChatError.invalidMessage }
        if message.kind != .text, message.attachment == nil { throw EncryptedChatError.invalidMessage }
        var output = magic
        output.append(1)
        output.append(message.kind.rawValue)
        output.append(contentsOf: uuidBytes(message.messageId))
        output.append(contentsOf: uuidBytes(message.channelId))
        append(message.membershipEpoch, to: &output)
        append(Int64(message.sentAt.timeIntervalSince1970 * 1_000), to: &output)
        append(UInt32(text.count), to: &output)
        output.append(text)
        guard let attachment = message.attachment else { return output }
        let name = Data(attachment.fileName.utf8)
        let mime = Data(attachment.mimeType.utf8)
        guard !name.isEmpty, name.count <= 255, !mime.isEmpty, mime.count <= 127,
              (1...Int64(maximumAttachmentBytes)).contains(attachment.plaintextBytes),
              (0...600_000).contains(attachment.durationMs), attachment.key.count == 32,
              attachment.ciphertextSha256.count == 32 else { throw EncryptedChatError.invalidAttachment }
        output.append(contentsOf: uuidBytes(attachment.attachmentId))
        append(attachment.plaintextBytes, to: &output)
        append(attachment.durationMs, to: &output)
        output.append(UInt8(name.count))
        output.append(UInt8(mime.count))
        output.append(attachment.key)
        output.append(attachment.ciphertextSha256)
        output.append(name)
        output.append(mime)
        return output
    }

    public static func decode(
        _ bytes: Data,
        senderAci: String,
        senderDeviceId: Int
    ) throws -> ChatMessage {
        guard bytes.count >= 54, bytes.prefix(4) == magic, bytes[4] == 1,
              let kind = ChatContentKind(rawValue: bytes[5]), (1...2).contains(senderDeviceId),
              UUID(uuidString: senderAci) != nil else { throw EncryptedChatError.invalidMessage }
        var offset = 6
        let messageId = try readUUID(bytes, &offset)
        let channelId = try readUUID(bytes, &offset)
        let epoch: Int32 = try read(bytes, &offset)
        let sentAtMs: Int64 = try read(bytes, &offset)
        let textCount = Int(try read(bytes, &offset) as UInt32)
        guard epoch > 0, textCount <= maximumTextBytes, offset + textCount <= bytes.count else {
            throw EncryptedChatError.invalidMessage
        }
        guard let text = String(data: bytes.subdata(in: offset..<(offset + textCount)), encoding: .utf8) else {
            throw EncryptedChatError.invalidMessage
        }
        offset += textCount
        var attachment: ChatAttachment?
        if kind == .text {
            guard offset == bytes.count else { throw EncryptedChatError.invalidMessage }
        } else {
            guard offset + 16 + 8 + 4 + 2 + 64 <= bytes.count else { throw EncryptedChatError.invalidAttachment }
            let attachmentId = try readUUID(bytes, &offset)
            let plaintextBytes: Int64 = try read(bytes, &offset)
            let durationMs: Int32 = try read(bytes, &offset)
            let nameCount = Int(bytes[offset]); offset += 1
            let mimeCount = Int(bytes[offset]); offset += 1
            let key = bytes.subdata(in: offset..<(offset + 32)); offset += 32
            let digest = bytes.subdata(in: offset..<(offset + 32)); offset += 32
            guard nameCount > 0, mimeCount > 0, offset + nameCount + mimeCount == bytes.count,
                  (1...Int64(maximumAttachmentBytes)).contains(plaintextBytes),
                  (0...600_000).contains(durationMs),
                  let name = String(data: bytes.subdata(in: offset..<(offset + nameCount)), encoding: .utf8)
            else { throw EncryptedChatError.invalidAttachment }
            offset += nameCount
            guard let mime = String(data: bytes.subdata(in: offset..<(offset + mimeCount)), encoding: .utf8) else {
                throw EncryptedChatError.invalidAttachment
            }
            attachment = ChatAttachment(
                attachmentId: attachmentId,
                fileName: name,
                mimeType: mime,
                plaintextBytes: plaintextBytes,
                durationMs: durationMs,
                key: key,
                ciphertextSha256: digest
            )
        }
        return ChatMessage(
            messageId: messageId,
            channelId: channelId,
            membershipEpoch: epoch,
            sentAt: Date(timeIntervalSince1970: TimeInterval(sentAtMs) / 1_000),
            senderAci: senderAci.lowercased(),
            senderDeviceId: senderDeviceId,
            kind: kind,
            text: text,
            attachment: attachment
        )
    }

    public static func sealAttachment(
        _ plaintext: Data,
        attachmentId: UUID,
        channelId: UUID,
        membershipEpoch: Int32,
        key: Data = .random(count: 32)
    ) throws -> (ciphertext: Data, key: Data, sha256: Data) {
        guard !plaintext.isEmpty, plaintext.count <= maximumAttachmentBytes, key.count == 32,
              membershipEpoch > 0 else { throw EncryptedChatError.invalidAttachment }
        let sealed = try AES.GCM.seal(
            plaintext,
            using: SymmetricKey(data: key),
            authenticating: attachmentAad(attachmentId, channelId, membershipEpoch)
        )
        guard let combined = sealed.combined else { throw EncryptedChatError.invalidAttachment }
        var ciphertext = attachmentMagic
        ciphertext.append(1)
        ciphertext.append(combined)
        return (ciphertext, key, Data(SHA256.hash(data: ciphertext)))
    }

    public static func openAttachment(
        _ ciphertext: Data,
        metadata: ChatAttachment,
        channelId: UUID,
        membershipEpoch: Int32
    ) throws -> Data {
        guard ciphertext.count > 5, ciphertext.prefix(4) == attachmentMagic, ciphertext[4] == 1,
              Data(SHA256.hash(data: ciphertext)) == metadata.ciphertextSha256 else {
            throw EncryptedChatError.attachmentIntegrityFailed
        }
        let box = try AES.GCM.SealedBox(combined: ciphertext.dropFirst(5))
        let plaintext = try AES.GCM.open(
            box,
            using: SymmetricKey(data: metadata.key),
            authenticating: attachmentAad(metadata.attachmentId, channelId, membershipEpoch)
        )
        guard plaintext.count == metadata.plaintextBytes else { throw EncryptedChatError.attachmentIntegrityFailed }
        return plaintext
    }

    private static func attachmentAad(_ attachmentId: UUID, _ channelId: UUID, _ epoch: Int32) -> Data {
        var value = Data("PTT-CHAT-ATTACHMENT-V1".utf8)
        value.append(contentsOf: uuidBytes(attachmentId))
        value.append(contentsOf: uuidBytes(channelId))
        append(epoch, to: &value)
        return value
    }

    private static func uuidBytes(_ value: UUID) -> [UInt8] {
        withUnsafeBytes(of: value.uuid) { Array($0) }
    }

    private static func readUUID(_ data: Data, _ offset: inout Int) throws -> UUID {
        guard offset + 16 <= data.count else { throw EncryptedChatError.invalidMessage }
        let bytes = [UInt8](data[offset..<(offset + 16)])
        offset += 16
        return UUID(uuid: (bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7],
                           bytes[8], bytes[9], bytes[10], bytes[11], bytes[12], bytes[13], bytes[14], bytes[15]))
    }

    private static func append<T: FixedWidthInteger>(_ value: T, to data: inout Data) {
        var big = value.bigEndian
        withUnsafeBytes(of: &big) { data.append(contentsOf: $0) }
    }

    private static func read<T: FixedWidthInteger>(_ data: Data, _ offset: inout Int) throws -> T {
        guard offset + MemoryLayout<T>.size <= data.count else { throw EncryptedChatError.invalidMessage }
        var value: T = 0
        _ = withUnsafeMutableBytes(of: &value) { data.copyBytes(to: $0, from: offset..<(offset + MemoryLayout<T>.size)) }
        offset += MemoryLayout<T>.size
        return T(bigEndian: value)
    }
}

public enum EncryptedChatError: Error, Equatable {
    case invalidMessage
    case textTooLarge
    case invalidAttachment
    case attachmentIntegrityFailed
}

File: android/talk/src/main/kotlin/app/ptt/talk/EncryptedChat.kt
package app.ptt.talk

import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.security.MessageDigest
import java.security.SecureRandom
import java.time.Instant
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

internal enum class ChatContentKind(val wire: Byte) { TEXT(1), FILE(2), VOICE(3), VIDEO(4) }

internal data class ChatAttachment(
    val attachmentId: UUID,
    val fileName: String,
    val mimeType: String,
    val plaintextBytes: Long,
    val durationMs: Int = 0,
    val key: ByteArray,
    val ciphertextSha256: ByteArray,
)

internal data class ChatMessage(
    val messageId: UUID,
    val channelId: UUID,
    val membershipEpoch: Int,
    val sentAt: Instant,
    val senderAci: String,
    val senderDeviceId: Int,
    val kind: ChatContentKind,
    val text: String,
    val attachment: ChatAttachment? = null,
)

internal object EncryptedChatCodec {
    const val MAX_TEXT_BYTES = 4_096
    const val MAX_ATTACHMENT_BYTES = 25 * 1_024 * 1_024
    private val MAGIC = "PTTC".encodeToByteArray()
    private val ATTACHMENT_MAGIC = "PTTA".encodeToByteArray()

    fun boundedUtf8(value: String, maximumBytes: Int): String {
        require(maximumBytes > 0)
        if (value.encodeToByteArray().size <= maximumBytes) return value
        val result = StringBuilder()
        val codePoints = value.codePoints().iterator()
        while (codePoints.hasNext()) {
            val codePoint = codePoints.nextInt()
            val candidate = result.toString() + String(Character.toChars(codePoint))
            if (candidate.encodeToByteArray().size > maximumBytes) break
            result.appendCodePoint(codePoint)
        }
        return result.toString()
    }

    fun encode(message: ChatMessage): ByteArray {
        require(message.membershipEpoch > 0 && message.senderDeviceId in 1..2)
        UUID.fromString(message.senderAci)
        val text = message.text.encodeToByteArray()
        require(text.size <= MAX_TEXT_BYTES)
        require((message.kind == ChatContentKind.TEXT) == (message.attachment == null))
        val attachment = message.attachment
        val extra = if (attachment == null) 0 else {
            val name = attachment.fileName.encodeToByteArray()
            val mime = attachment.mimeType.encodeToByteArray()
            require(name.size in 1..255 && mime.size in 1..127)
            require(attachment.plaintextBytes in 1..MAX_ATTACHMENT_BYTES.toLong())
            require(attachment.durationMs in 0..600_000 && attachment.key.size == 32 && attachment.ciphertextSha256.size == 32)
            16 + 8 + 4 + 2 + 64 + name.size + mime.size
        }
        return ByteBuffer.allocate(4 + 1 + 1 + 16 + 16 + 4 + 8 + 4 + text.size + extra).apply {
            put(MAGIC).put(1).put(message.kind.wire)
            putUuid(message.messageId).putUuid(message.channelId)
            putInt(message.membershipEpoch).putLong(message.sentAt.toEpochMilli()).putInt(text.size).put(text)
            if (attachment != null) {
                val name = attachment.fileName.encodeToByteArray()
                val mime = attachment.mimeType.encodeToByteArray()
                putUuid(attachment.attachmentId).putLong(attachment.plaintextBytes).putInt(attachment.durationMs)
                put(name.size.toByte()).put(mime.size.toByte()).put(attachment.key).put(attachment.ciphertextSha256)
                put(name).put(mime)
            }
        }.array()
    }

    fun decode(bytes: ByteArray, senderAci: String, senderDeviceId: Int): ChatMessage {
        require(bytes.size >= 54 && senderDeviceId in 1..2)
        UUID.fromString(senderAci)
        val buffer = ByteBuffer.wrap(bytes)
        require(ByteArray(4).also(buffer::get).contentEquals(MAGIC) && buffer.get().toInt() == 1)
        val kindWire = buffer.get()
        val kind = ChatContentKind.entries.firstOrNull { it.wire == kindWire } ?: error("invalid chat kind")
        val messageId = buffer.uuid()
        val channelId = buffer.uuid()
        val epoch = buffer.int
        val sentAt = Instant.ofEpochMilli(buffer.long)
        val textLength = buffer.int
        require(epoch > 0 && textLength in 0..MAX_TEXT_BYTES && textLength <= buffer.remaining())
        val text = strictUtf8(ByteArray(textLength).also(buffer::get))
        val attachment = if (kind == ChatContentKind.TEXT) {
            require(!buffer.hasRemaining())
            null
        } else {
            require(buffer.remaining() >= 16 + 8 + 4 + 2 + 64)
            val id = buffer.uuid()
            val size = buffer.long
            val duration = buffer.int
            val nameLength = buffer.get().toInt() and 0xff
            val mimeLength = buffer.get().toInt() and 0xff
            val key = ByteArray(32).also(buffer::get)
            val digest = ByteArray(32).also(buffer::get)
            require(nameLength > 0 && mimeLength > 0 && nameLength + mimeLength == buffer.remaining())
            require(size in 1..MAX_ATTACHMENT_BYTES.toLong() && duration in 0..600_000)
            val name = strictUtf8(ByteArray(nameLength).also(buffer::get))
            val mime = strictUtf8(ByteArray(mimeLength).also(buffer::get))
            ChatAttachment(id, name, mime, size, duration, key, digest)
        }
        return ChatMessage(messageId, channelId, epoch, sentAt, senderAci.lowercase(), senderDeviceId, kind, text, attachment)
    }

    fun sealAttachment(
        plaintext: ByteArray,
        attachmentId: UUID,
        channelId: UUID,
        membershipEpoch: Int,
        key: ByteArray = ByteArray(32).also(SecureRandom()::nextBytes),
    ): Triple<ByteArray, ByteArray, ByteArray> {
        require(plaintext.size in 1..MAX_ATTACHMENT_BYTES && key.size == 32 && membershipEpoch > 0)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val nonce = ByteArray(12).also(SecureRandom()::nextBytes)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(attachmentAad(attachmentId, channelId, membershipEpoch))
        val ciphertext = ATTACHMENT_MAGIC + byteArrayOf(1) + nonce + cipher.doFinal(plaintext)
        return Triple(ciphertext, key.copyOf(), MessageDigest.getInstance("SHA-256").digest(ciphertext))
    }

    fun openAttachment(ciphertext: ByteArray, metadata: ChatAttachment, channelId: UUID, membershipEpoch: Int): ByteArray {
        require(ciphertext.size > 5 + 12 + 16 && ciphertext.copyOfRange(0, 4).contentEquals(ATTACHMENT_MAGIC) && ciphertext[4].toInt() == 1)
        require(MessageDigest.getInstance("SHA-256").digest(ciphertext).contentEquals(metadata.ciphertextSha256))
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(metadata.key, "AES"), GCMParameterSpec(128, ciphertext.copyOfRange(5, 17)))
        cipher.updateAAD(attachmentAad(metadata.attachmentId, channelId, membershipEpoch))
        return cipher.doFinal(ciphertext, 17, ciphertext.size - 17).also {
            require(it.size.toLong() == metadata.plaintextBytes)
        }
    }

    private fun attachmentAad(attachmentId: UUID, channelId: UUID, membershipEpoch: Int): ByteArray =
        ByteBuffer.allocate(22 + 16 + 16 + 4).put("PTT-CHAT-ATTACHMENT-V1".encodeToByteArray())
            .putUuid(attachmentId).putUuid(channelId).putInt(membershipEpoch).array()

    private fun strictUtf8(bytes: ByteArray): String =
        Charsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT).decode(ByteBuffer.wrap(bytes)).toString()

    private fun ByteBuffer.putUuid(value: UUID): ByteBuffer = putLong(value.mostSignificantBits).putLong(value.leastSignificantBits)
    private fun ByteBuffer.uuid(): UUID = UUID(long, long)
}
